### PR TITLE
Refactor OsuDifficultyHitObject for readability/understandability

### PR DESCRIPTION
**Pull request by smoogipoo**
I recommend going through this PR per-commit as the diff is a little bit of a mess.

- The primary issue I had is that `TravelDistance` and `TravelTime` stored information about the _previous_ hitobject, in the current hitobject. I've made these two store information about the current hitobject only.
- The rest is a bit of refactoring to clean things up - property ordering, naming (American vs British English), and expanding on the comment describing how `MovementDistance` is calculated.

I've tested that SR and difficulty attributes don't change compared to current master.

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
Comment (peppy) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:50]: Confirm that this is still correct as `osuLastObj` here? Looks weird with the changes inside this conditional now reading from one index back while the conditional itself isn't.

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:50]: This is one of the reasons for me changing it, because it was using `osuCurrObj.TravelDistance` to access the travel distance of `osuLastObj`.

So yes, this condition is still fine as it is, the only change is that the properties apply directly to the relevant hitobject now.

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:50]: To add on as to why - all these blocks of code try to do (perhaps not explained enough in the leading comments?) is to calculate the velocity through a previous slider into the current object. One part is stored in the previous object (travel), and the other is stored in the current object (jump / movement, however you want to think about it).

**Review by peppy — DISMISSED** [review]

**Comment by smoogipoo** [comment]
Don't merge this yet as I realised my testing was flawed. Will unblock after re-confirming SR is unaffected.

**Comment by smoogipoo** [comment]
Found an issue and re-tested, looks okay now.

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
Comment (stanriders) [inline osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:37]: Maybe rename it to something like `MinimumMovementDistance`? Overall distance/time fields could use better naming so that it can be more obvious which does what without reading docs every time.

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:36]: I'm still not sure how to read this correctly. There's a mention of a "minimum distance from end position of previous object", but then how can that be anything *other* than `JumpDistance` if it's in a straight line?

The "end position" part I especially take issue with, because as far as I can tell from [the picture in the other PR thread correctly](https://github.com/ppy/osu/pull/15758#issuecomment-977468402) and some inline comments from this pull, the point is that a path with length of `MovementDistance` does *not* necessarily hit the end position of the previous object, it just moves the cursor enough to *not miss it*?

Given my understanding above is correct, only then it starts to makes sense to me that `JumpDistance` is the upper bound of `MovementDistance`. (I would also specify the "upper bound" part explicitly in this xmldoc.)

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:53]: Nit: now with this change the code reads back to front chronologically a bit (current object is used before last). It looks correct, but I'd just swap order to follow more naturally.

**Review by bdach — COMMENTED** [review]
Haven't looked at diffcalc for a while but the transformation generally looks correct to me. Just want to clarify one xmldoc (there's also a nitpick in this review but I don't care about it very much)

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:53]: Funny. I originally had inverted these lines because of the same feeling, but reverted in order to reduce changes. Agree.

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
Comment (bdach) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:36]: I'm not sure where this review conversation is at after the latest changes - is there something I should recommend here? Because I'm admittedly still confused as to what `Jump` and `MovementDistance` are

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:36]: I haven't gotten around to looking into it yet. The pushed commit was supposed to be pushed quite a long time ago, it was just sitting in my local working set.

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs]: Renamed to `MinimumJumpDistance`.

**Comment by stanriders** [comment]
Yep, seems much better to me now

**Comment by bdach** [comment]
The latest update to docs/names looks good to me now, but there is a merge conflict here and I'm unsure whether my resolution is correct... I *think* this would be the correct one?

```diff
diff --cc osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs
index 75a9b13bdf,4f87767fa7..ca89236258
--- a/osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs
+++ b/osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs
@@@ -154,8 -154,7 +154,7 @@@ namespace osu.Game.Rulesets.Osu.Difficu
              if (strainTime < min_speed_bonus)
                  speedBonus = 1 + 0.75 * Math.Pow((min_speed_bonus - strainTime) / speed_balancing_factor, 2);
  
-             double travelDistance = osuPrevObj?.TravelDistance ?? 0;
-             double distance = Math.Min(single_spacing_threshold, travelDistance + osuCurrObj.LazyJumpDistance);
 -            double distance = Math.Min(single_spacing_threshold, osuCurrObj.TravelDistance + osuCurrObj.MovementDistance);
++            double distance = Math.Min(single_spacing_threshold, osuCurrObj.TravelDistance + osuCurrObj.MinimumJumpDistance);
  
              return (speedBonus + speedBonus * Math.Pow(distance / single_spacing_threshold, 3.5)) / strainTime;
          }
```

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
Comment (peppy) [inline osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:131]: I'd recommend representing the cursor location with something that isn't a "o". This reads as if it's hitobjects, which can make for confusion here. How about using a `*` or `x` or something?

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs]: Sure.

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
Comment (peppy) [inline osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:126]: I still don't quite understand what this diagram is trying to explain, but as long as others can maybe it's okay.

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:126]: The diagrams show how the natural movement path of a player changes depending on what the angle of a follow-up circle is, and how using the lazy cursor end is completely fine in diagram 1 ("players will cut the slider short in order to move to the next object"), but will overestimate the perceived jump distance in diagram 2 ("players will follow the slider through to its visual extent").

I don't know how to better explain it.

Comment (peppy) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:126]: Am I supposed to know that there's a second object below the slider in the first diagram? So the "x" represents the position of the slider ball and also the position of the next object here?

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:126]: No, it represents the final position your cursor needs to be in a slider given the most lazy movements (following at the edge of the follow circle at all times) + the -36ms offset for the last slider tick.

Comment (peppy) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:126]: But the downwards movement implies there's something beneath it right? If so it's be nice to show that next object to give more context.

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:126]: I misunderstood what you meant, I thought you were referring to the "x" on the slider track.

How should I show that the "x" below the slider is another hitobject? Any suggestions? Would this be better?
image
I can't easily add the same text to the second diagram...

Comment (peppy) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:126]: Maybe they should be shown using `o` after all, and the documentation should not mention anything about the cursor position. The cursor movement is shown by the single-width line, which should be ample to explain the concept here.

Might help to split the diagram out into two versions beside each other: one with the "lazy" movement and one with the "correct" movement (ie the one denoted by `/`). Feels weird showing both on the same diagram.

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs]: I've restructured it, see if it's better now?

Comment (peppy) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs]: Yep, much better.

**Comment by smoogipoo** [comment]
@bdach I've resolved conflicts, note that your patch is slightly incorrect because this PR moves `TravelDistance` to the hitobject which is actually being traveled.

**Review by bdach — APPROVED** [review]

**Review by peppy — APPROVED** [review]

**Pull request merged**
